0.12.0 non-x86 compile failure

## 🐛 Bug Reports

When reporting a bug, please provide the following information. If this is not a bug report you can just discard this template.

### 🌍 Environment

 - Your operating system and version: Linux (ubuntu 18.04 on ppc64le, s390x, and aarch64
 - Your python version: 3.7
 - How did you install python (e.g. apt or pyenv)? Did you use a virtualenv?:
 - Your Rust version (`rustc --version`):
 - Your PyO3 version: 0.12.0
 - Have you tried using latest PyO3 master (replace `version = "0.x.y"` with `git = "https://github.com/PyO3/pyo3")?`: no

### 💥 Reproducing

Building the pyo3 0.12.0 crate on non-x86 linux seems to fail because of a typing issue for a raw pointer. I started hitting this in CI for my downstream project when trying to upgrade to 0.12.0 here:

https://travis-ci.com/github/Qiskit/retworkx/jobs/384362604#L448
https://travis-ci.com/github/Qiskit/retworkx/jobs/384362605#L453
https://travis-ci.com/github/Qiskit/retworkx/jobs/384362606#L455

It looks like this was changed in https://github.com/PyO3/pyo3/commit/b9e95dc7c92b09d07e3a1f4a54e65d6753eda548#diff-8a0564885baee0a1d53807fd6ef013e5R448 the underlying issue is that a `c_char` can either be an `i8` or `u8` https://doc.rust-lang.org/std/os/raw/type.c_char.html and it looks like there is a mismatch on non-x86 systems. 

Maybe we should just cast the input there as a `c_char` instead of `i8`?

(the error for when the travis log expires is below)
```
Running `rustc --crate-name pyo3 --edition=2018 /home/travis/.cargo/registry/src/github.com-1ecc6299db9ec823/pyo3-0.12.0/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi --crate-type lib --emit=dep-info,metadata,link -C opt-level=3 -Cembed-bitcode=no --cfg 'feature="ctor"' --cfg 'feature="default"' --cfg 'feature="extension-module"' --cfg 'feature="indoc"' --cfg 'feature="inventory"' --cfg 'feature="macros"' --cfg 'feature="paste"' --cfg 'feature="pyo3cls"' --cfg 'feature="unindent"' -C metadata=cf6285843d911853 -C extra-filename=-cf6285843d911853 --out-dir /tmp/pip-req-build-xiu1t4n9/target/release/deps -L dependency=/tmp/pip-req-build-xiu1t4n9/target/release/deps --extern ctor=/tmp/pip-req-build-xiu1t4n9/target/release/deps/libctor-d476a6cabbb1d5d0.so --extern indoc=/tmp/pip-req-build-xiu1t4n9/target/release/deps/libindoc-53e0ec533919e29f.rmeta --extern inventory=/tmp/pip-req-build-xiu1t4n9/target/release/deps/libinventory-68485ad54a661ce6.rmeta --extern libc=/tmp/pip-req-build-xiu1t4n9/target/release/deps/liblibc-76db962b748c2a37.rmeta --extern parking_lot=/tmp/pip-req-build-xiu1t4n9/target/release/deps/libparking_lot-ae5080acef9548d0.rmeta --extern paste=/tmp/pip-req-build-xiu1t4n9/target/release/deps/libpaste-be221bde5cca2a82.rmeta --extern pyo3cls=/tmp/pip-req-build-xiu1t4n9/target/release/deps/libpyo3cls-5dbf26b63a171d5e.so --extern unindent=/tmp/pip-req-build-xiu1t4n9/target/release/deps/libunindent-fa42d75d448dc224.rmeta --cap-lints allow --cfg Py_3_5 --cfg Py_3_6 --cfg Py_3_7 --cfg Py_3 --cfg 'py_sys_config="WITH_THREAD"'`

  error[E0308]: mismatched types

     --> /home/travis/.cargo/registry/src/github.com-1ecc6299db9ec823/pyo3-0.12.0/src/exceptions.rs:448:17

      |

  448 |                 input.as_ptr() as *const i8,

      |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `u8`, found `i8`

      |

      = note: expected raw pointer `*const u8`

                 found raw pointer `*const i8`

  

  error: aborting due to previous error

  

  For more information about this error, try `rustc --explain E0308`.

  error: could not compile `pyo3`.

  ```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -11,6 +11,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 ### Removed
 ### Fixed
 - Fix building for a 32-bit Python on 64-bit Windows with a 64-bit Rust toolchain. [#1179](https://github.com/PyO3/pyo3/pull/1179)
+- Fix building on platforms where `c_char` is `u8` [#1182][https://github.com/PyO3/pyo3/pull/1182]
 
 ## [0.12.0] - 2020-09-12
 ### Added

--- a/src/exceptions.rs
+++ b/src/exceptions.rs
@@ -6,6 +6,7 @@ use crate::type_object::PySizedLayout;
 use crate::{ffi, PyResult, Python};
 use std::ffi::CStr;
 use std::ops;
+use std::os::raw::c_char;
 
 /// The boilerplate to convert between a Rust type and a Python exception.
 #[macro_export]
@@ -445,7 +446,7 @@ impl PyUnicodeDecodeError {
         unsafe {
             py.from_owned_ptr_or_err(ffi::PyUnicodeDecodeError_Create(
                 encoding.as_ptr(),
-                input.as_ptr() as *const i8,
+                input.as_ptr() as *const c_char,
                 input.len() as ffi::Py_ssize_t,
                 range.start as ffi::Py_ssize_t,
                 range.end as ffi::Py_ssize_t,

